core/app/models/spree/shipment.rb

```diff
@@ -130,10 +130,17 @@ def final_price_with_items
       item_cost + final_price
     end
 
+    # Decrement the stock counts for all pending inventory units in this
+    # shipment and mark.
+    # Any previous non-pending inventory units are skipped as their stock had
+    # already been allocated.
     def finalize!
-      return if order.unreturned_exchange?
-      InventoryUnit.finalize_units!(inventory_units)
-      manifest.each { |item| manifest_unstock(item) }
+      transaction do
+        pending_units = inventory_units.select(&:pending?)
+        pending_manifest = ShippingManifest.new(inventory_units: pending_units)
+        pending_manifest.items.each { |item| manifest_unstock(item) }
```

Comment:
The inventory units are based on a line items quantity right? I might use a find each here in case people use Solidus for wholesale of goods; but now I'm just trolling.